Data rate fixes

spiiroin opened the pull request:
[nodebase] Use closest supported data rate. Fixes JB#60314

What kinds of data rates are available for sensors varies from one
device to another. Having sensorfwd completely ignore out of bounds
data rate requests creates a situation where applications can work in
completely different ways on different devices.

Instead of rejecting out of bounds data requests, choose closest
supported data rate.

[sensorfw] Drop evaluateIntervalRequests overloading. JB#60314

NodeBase::evaluateIntervalRequests is virtual method that about every
derived class overrides - with almost identical copy paste code.

The only difference between these variants is how placeholder zero
interval is handled. Base behavior seems to want to treat it as "as
fast as possible" while others take it to mean "use the defaults."

When intervals were defined in millisecond granularity, using zero
might have had some use for reaching > 1kHz data rates. Now that
microseconds are used, the boundary for expressable data rates has
moved to >1MHz level and dual role for zero value is less useful.

Change NodeBase behavior to match what the multitude of derived
classes used to do and drop all the now unnecessary copy-paste
methods.

[hybrisadaptor] Fix setDelay() return value

If data rate adjustment is skipped because data rate did not change,
setDelay() returns failure - which then causes journal spamming and
other hiccups.

Return success if data rate adjustment is skipped on purpose.

[sensorfw] Handle intervals in microsecond granularity. Fixes JB#60313

Sample intervals within sensorfwd are handled in microsecond granularity
and converted to appropriate units when dealing with external interfaces
such as D-Bus, Android hal, or sysfs control files.

[sensorfw] Add setDataRate() D-Bus method. JB#60417

Using integer millisecond sampling interval is problematic for
specifying high acquisition rates as it limits choice to
1000, 500, 333, 250, 200, ... Hz and thus rules out e.g. 400 Hz
altogether.

Use microsecond granularity interval time internally and provide new
setDataRate() method that takes double precision floating point
Hertz parameter - while retaining the already existing setInterval()
method for the sake of backwards compatibility.

Note that this only deals with D-Bus interface level. Further changes
in sensorfw side are needed to actually support e.g. 400 Hz data rates.

[sensorfw] Explicitly indicate interval time units. JB#60313

As a preparatory step for using different time units in D-Bus interface
and internally within sensorfw: use "_ms" suffix for all interval
arguments and member variables.

Also make affected classes follow m_variableName naming convention.

[sensorfw] Normalize setInterval() parameter order. Fixes JB#60301

The order of interval and session id parameters for setInterval() methods
varies within class hierarchy present in sensorfw.

Normalize situation so that when setInterval() method takes session id
parameter, it is always the first argument to pass - which aligns with
dbus interface.

spiiroin committed:
[sensorfw] Normalize setInterval() parameter order. Fixes JB#60301

The order of interval and session id parameters for setInterval() methods
varies within class hierarchy present in sensorfw.

Normalize situation so that when setInterval() method takes session id
parameter, it is always the first argument to pass - which aligns with
dbus interface.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

spiiroin committed:
[sensorfw] Explicitly indicate interval time units. JB#60313

As a preparatory step for using different time units in D-Bus interface
and internally within sensorfw: use "_ms" suffix for all interval
arguments and member variables.

Also make affected classes follow m_variableName naming convention.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

spiiroin committed:
[sensorfw] Add setDataRate() D-Bus method. JB#60417

Using integer millisecond sampling interval is problematic for
specifying high acquisition rates as it limits choice to
1000, 500, 333, 250, 200, ... Hz and thus rules out e.g. 400 Hz
altogether.

Use microsecond granularity interval time internally and provide new
setDataRate() method that takes double precision floating point
Hertz parameter - while retaining the already existing setInterval()
method for the sake of backwards compatibility.

Note that this only deals with D-Bus interface level. Further changes
in sensorfw side are needed to actually support e.g. 400 Hz data rates.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

spiiroin committed:
[sensorfw] Handle intervals in microsecond granularity. Fixes JB#60313

Sample intervals within sensorfwd are handled in microsecond granularity
and converted to appropriate units when dealing with external interfaces
such as D-Bus, Android hal, sysfs controls, or configuration files.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

spiiroin committed:
[sensorfw] Drop evaluateIntervalRequests overloading. JB#60314

NodeBase::evaluateIntervalRequests is virtual method that about every
derived class overrides - with almost identical copy paste code.

The only difference between these variants is how placeholder zero
interval is handled. Base behavior seems to want to treat it as "as
fast as possible" while others take it to mean "use the defaults."

When intervals were defined in millisecond granularity, using zero
might have had some use for reaching > 1kHz data rates. Now that
microseconds are used, the boundary for expressable data rates has
moved to >1MHz level and dual role for zero value is less useful.

Change NodeBase behavior to match what the multitude of derived
classes used to do and drop all the now unnecessary copy-paste
methods.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

spiiroin committed:
[hybrisadaptor] Fix setDelay() return value

If data rate adjustment is skipped because data rate did not change,
setDelay() returns failure - which then causes journal spamming and
other hiccups.

Return success if data rate adjustment is skipped on purpose.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

mlehtima commented on core/nodebase.cpp:
```diff
@@ -427,42 +428,42 @@ void NodeBase::setIntervalSource(NodeBase* node)
+bool NodeBase::setDefaultInterval(const unsigned int interval_us)
 {
-    if (!isValidIntervalRequest(value))
+    unsigned validatedInterval_ms = validateIntervalRequest(interval_us);
```
Should be validatedInterval_us instead of validatedInterval_ms to match the units of input.

spiiroin replied:
yes. also it seems that I had missed default interval setup altogether -> will add separate commit about that

mlehtima commented on core/nodebase.cpp:
```diff
@@ -314,23 +314,21 @@ unsigned int NodeBase::getInterval(int sessionId) const
-        sensordLogW() << "Invalid interval requested for node '" << id() << "' by session '" << sessionId << "': " << value;
-        return false;
-    }
+    unsigned validatedInterval_ms = validateIntervalRequest(interval_us);
```
Should be validatedInterval_us instead of validatedInterval_ms to match the units of input. Also lets use unsigned int instead of just unsigned for consistency. This applies to other places also.

spiiroin replied:
all (new) occurenses of 'unsigned' should now be changed to 'unsigned int'

mlehtima commented on core/abstractsensor_a.h:
```diff
@@ -124,7 +124,13 @@ public Q_SLOTS: // METHODS
+     *
+     *  Will also configure interval for the data connection.
+     */
+    void setDataRate(int sessionId, double value);
```
Would it make sense to name the parameter as dataRate_Hz the same way it is in .cpp file?

spiiroin replied:
yes, that was the intent

mlehtima commented on qt-api/abstractsensor_i.cpp:
```diff
@@ -369,57 +395,85 @@ QString AbstractSensorChannelInterface::id()
+double AbstractSensorChannelInterface::dataRate()
+{
+    double dataRate_Hz = 0;
+    if( pimpl_->m_interval_us > 0)
```
Inconsistent formatting, "if (" is used elsewhere in the file.

spiiroin replied:
ack, will fix. did not find others

mlehtima commented on adaptors/iioadaptor/iioadaptor.cpp:
```diff
@@ -548,21 +548,22 @@ void IioAdaptor::processSample(int fileId, int fd)
-//    int value = 100;
-//    sensordLogD() << "Returning dummy value in interval(): " << value;
-//    return value;
+//    int interval_us = 100;
```
Probably would make sense the change also the commented out value to match new units in case this gets enabled at some point.

spiiroin replied:
yes, will fix.

[in general this kind of  dead code updating feels a bit stupid, but... maybe some of them might be relevant some day]

mlehtima commented on tests/testapp/abstractsensorhandler.cpp:
```diff
@@ -29,22 +29,22 @@
-    downsample_(false)
+    m_sensorName(sensorName),
+    m_interval_ms(100),
+    m_bufferinterval(0),
```
MIght be more clear if this also had unit in variable name like other places already have.

spiiroin replied:
another forgotten subitem... buffer intervals might be for sw timers which would need to be millisecs at use time. need to take a closer look.

mlehtima commented on core/nodebase.cpp:
```diff
@@ -330,7 +330,7 @@ bool NodeBase::setIntervalRequest(const int sessionId, const unsigned int interv
 
     // Store the request for the session
-    m_intervalMap[sessionId] = interval_us;
+    m_intervalMap[sessionId] = validatedInterval_ms;
```
This change should be in next commit "Use closest supported data rate". Also remove the extra empty line change below from this commit.

spiiroin replied:
yes, completely missed this, will fix

spiiroin committed:
[nodebase] Use closest supported data rate. Fixes JB#60314

What kinds of data rates are available for sensors varies from one
device to another. Having sensorfwd completely ignore out of bounds
data rate requests creates a situation where applications can work in
completely different ways on different devices.

Instead of rejecting out of bounds data requests, choose closest
supported data rate.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

spiiroin committed:
[declinationfilter] Add unit suffixes to time values

Declination deals with sensor event timestamps in microseconds.

No code changes required, but rename member variables to
conform with m_variableName / s_staticVariable convention
and add time unit suffixer.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

spiiroin committed:
[nodebase] Remove dead code

There is code for tracking buffer sizes and intervals in sessionId
based hash tables elsewhere - presumably these node base member
variables are left overs from time before those were taken in use?

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

spiiroin committed:
[nodebase] Use m_variableName convention for all members

For one or another reason both variableName_ and m_variableName
conventions are used in the same class.

Normalize to m_variableName style.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>

mlehtima commented on core/nodebase.cpp:
```diff
@@ -430,27 +430,24 @@ void NodeBase::setIntervalSource(NodeBase* node)
     }
+    if (chosenInterval_us == 0)
+        chosenInterval_us = defaultInterval();
+
```
This empty line is removed in next commit.

spiiroin replied:
removed

mlehtima commented on core/nodebase.cpp:
```diff
@@ -430,27 +430,24 @@ void NodeBase::setIntervalSource(NodeBase* node)
+    for (QMap<int, unsigned int>::const_iterator it = m_intervalMap.constBegin(); it != m_intervalMap.constEnd(); ++it) {
+        int iterSessionId = it.key();
+        unsigned int iterInterval_us = it.value();
+        if((iterInterval_us > 0 && iterInterval_us < chosenInterval_us) || (chosenInterval_us == 0)) {
```
Add space after if.

spiiroin replied:
added
